Autoscaling APIs not experimental

henningandersen opened the pull request:
We no longer regard the autoscaling APIs experimental though they are
only intended for use by ESS/ECE/ECK.  This commit updates the docs
to reflect this and adds a minimal set of documentation for the
feature.

henningandersen committed:
Autoscaling APIs not experimental

We no longer regard the autoscaling APIs experimental though they are
only intended for use by ESS/ECE/ECK. This commit updates the docs
to reflect this and adds a minimal set of documentation for the
feature.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

henningandersen commented on docs/reference/autoscaling/index.asciidoc:
```diff
@@ -1,10 +1,9 @@
-[testenv="platinum"]
+[testenv="enterprise"]
 [[xpack-autoscaling]]
-[chapter]
```
My asciidoc skills failed me here - removing this made this work, otherwise the sub page would appear on the same level in the navigation.

jrodewig commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+====
+`required_capacity`::
+(object)
+Contains the required capacity for the policy
```
Add period.
```suggestion
Contains the required capacity for the policy.
```

jrodewig commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+=====
+`node`::
+(object)
+Contains the minimum node sizes required per node (ensuring that individual
+shards or ML jobs can fit into a single node)
```
Minor punctuation changes.
```suggestion
Contains the minimum node sizes required per node, ensuring that individual
shards or ML jobs can fit into a single node.
```

jrodewig commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+=====
+`node`::
+(object)
+Contains the maximum sizes of nodes governed by the policy
```
```suggestion
Contains the maximum sizes of nodes governed by the policy.
```

jrodewig commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+======
+`<decider_name>`::
+(object)
+The capacity result for a specific decider that is enabled for the policy
```
```suggestion
The capacity result for a specific decider enabled for the policy.
```

jrodewig commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+======
+`total`::
+(object)
+Contains the total size required for the policy
```
Adds period.
```suggestion
Contains the total size required for the policy.
```

jrodewig commented on docs/reference/autoscaling/deciders/proactive-storage-decider.asciidoc:
```diff
@@ -0,0 +1,54 @@
+the current data set plus an estimated amount of expected additional data.
+
+It is enabled for all policies governing nodes with the `data_hot` role. A
+single configuration option `forecast_window` is available, defaulting to 30
+minutes.
+
+The estimation of expected additional data is based on past indexing that
+occurred within the `forecast_window`, i.e., by default the last 30 minutes.
+Only indexing into data streams contributes to the estimate.
```
We'd typically put this in a `Parameters` or `Options` section with a def for `forecast_window`.

henningandersen replied:
Did this in d0429bd4c243159e1571238588e123dde1eeedf4

jrodewig reviewed (approved):
This looks good overall. I left several suggestions, but feel free to ignore those if wanted.

Two larger pieces of non-blocking feedback:

* I'd add a sentence or two about what deciders are and how autoscaling uses them to the top-level page. There isn't a conceptual intro to deciders in the docs currently.
* I'd create a separate `Parameters` or `Options` section for configuration options in each decider reference page.

henningandersen committed:
Update docs/reference/autoscaling/index.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/index.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/deciders/reactive-storage-decider.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/deciders/reactive-storage-decider.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/deciders/proactive-storage-decider.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/deciders/proactive-storage-decider.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Update docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen commented on docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc:
```diff
@@ -0,0 +1,57 @@
+It is enabled for policies governing `ml` nodes and has following
+configuration options:
+
+* `num_anomaly_jobs_in_queue`, default 0
+* `num_analytics_jobs_in_queue`, default 0
+* `down_scale_delay`, default 1 hour
```
I will re-arrange and then ask ML team to add more detailed definitions in a follow-up.

henningandersen replied:
See d0429bd4c243159e1571238588e123dde1eeedf4

henningandersen committed:
Add outer policies

henningandersen committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

henningandersen committed:
Max bytes of storage/memory of a node.

henningandersen committed:
Merge branch 'docs_autoscaling_not_experimental' of github.com:henningandersen/elasticsearch into docs_autoscaling_not_experimental

henningandersen committed:
Designed for note.

henningandersen committed:
Add short intro to policy, roles and deciders.

henningandersen committed:
Add configuration settings sections.

henningandersen commented on docs/reference/autoscaling/autoscaling-designed-for-note.asciidoc:
```diff
@@ -0,0 +1,2 @@
+[NOTE]
+Autoscaling is designed for indirect use by ECE/ESS and ECK. Direct use is not supported.
```
I added the links in b8c73509249268858134b21484553dee465e7738, but kept `indirect` rather than `internal`, since ESS/ECE/ECK users will be able to configure the decider settings indirectly, using the setting names documented here through those products.

henningandersen commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+
+`<policy_name>`::
+(object)
+Contains capacity for the policy
```
Yes, thanks for spotting that, added this in 9bcf5bb60fdd165ced2ba0995fa3ffc2a3731352

henningandersen commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+======
+====
+
+
```
Let me handle that in a follow-up, is a bit tricky to get right.

henningandersen commented on docs/reference/autoscaling/index.asciidoc:
```diff
@@ -17,3 +16,11 @@ allocate existing and future shard copies due to disk space, then the
 autoscaling API would report that the cluster needs to scale due to disk space.
 It remains the responsibility of the operator to add the additional resources
 that the cluster signals it requires.
+
```
I added this in 7ab0ebfc6d40d28cd535ae7c3d10bb3123c3ebee

henningandersen commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+======
+`storage`::
+(integer)
+Contains the maximum current storage of a node, in bytes
```
I did half of that in 94a0a21d36cb55f379a9bb44138c01a4fe89973d, but kept "of a node". It is the max number of bytes across all nodes.

henningandersen commented on docs/reference/autoscaling/apis/get-autoscaling-capacity.asciidoc:
```diff
@@ -33,6 +33,189 @@ This API gets the current autoscaling capacity based on the configured
+
+`memory`::
+(integer)
+Contains the maximum current memory of a node, in bytes
```
I did half of that in 94a0a21d36cb55f379a9bb44138c01a4fe89973d, but kept "of a node". It is the max number of bytes across all nodes.

henningandersen commented:
Thanks so much for the many suggestions, @jrodewig. I believe I addressed all of them. I added a `Configuration settings` section to the decider pages in d0429bd4c243159e1571238588e123dde1eeedf4.

jrodewig commented on docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc:
```diff
@@ -0,0 +1,63 @@
+`num_analytics_jobs_in_queue`::
+Number of queued analytics jobs to allow. Default 0.
+`down_scale_delay`::
+(Optional, <<time-units,time value>>
```
```suggestion
(Optional, <<time-units,time value>>)
```

jrodewig commented on docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc:
```diff
@@ -0,0 +1,63 @@
+(Optional, integer)
+Number of queued anomaly jobs to allow. Default 0.
+
+`num_analytics_jobs_in_queue`::
+Number of queued analytics jobs to allow. Default 0.
```
```suggestion
`num_analytics_jobs_in_queue`::
(Optional, integer)
Number of queued analytics jobs to allow. Defaults to `0`.

```

jrodewig commented on docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc:
```diff
@@ -0,0 +1,63 @@
+Number of queued analytics jobs to allow. Default 0.
+`down_scale_delay`::
+(Optional, <<time-units,time value>>
+Delay before scaling down. Default 1 hour.
```
```suggestion
Delay before scaling down. Defaults to 1 hour.
```

jrodewig commented on docs/reference/autoscaling/deciders/machine-learning-decider.asciidoc:
```diff
@@ -0,0 +1,63 @@
+
+`num_anomaly_jobs_in_queue`::
+(Optional, integer)
+Number of queued anomaly jobs to allow. Default 0.
```
```suggestion
Number of queued anomaly jobs to allow. Defaults to `0`.
```

jrodewig reviewed (approved):
I left a few formatting nits, but this looks great to me. Thanks @henningandersen!

henningandersen committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

jasontedor reviewed (approved):
LGTM!

jasontedor commented on docs/reference/autoscaling/autoscaling-designed-for-note.asciidoc:
```diff
@@ -0,0 +1,2 @@
+[NOTE]
+Autoscaling is designed for indirect use by {ess-trial}[{ess}], {ece-ref}[{ece}], and {eck-ref}[Elastic Cloud on Kubernetes]. Direct use is not supported.
```
Minor nit:
```suggestion
Autoscaling is designed for indirect use by {ess-trial}[{ess}],
{ece-ref}[{ece}], and {eck-ref}[Elastic Cloud on Kubernetes]. Direct use is not
supported.
```

jasontedor replied:
Since it otherwise exceeds 80 columns.

henningandersen committed:
Update docs/reference/autoscaling/autoscaling-designed-for-note.asciidoc

Co-authored-by: Jason Tedor <jason@tedor.me>
